### ISSUE TITLE
RT-DETR-l training validation works but manual validation returns 0

### ISSUE THREAD

**Issue by enescicek** [ultralytics/ultralytics#24580]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I trained both YOLOv8l and RT-DETR-l models using the same dataset and the same training parameters.

After training, I ran validation on the YOLOv8l model without any issues and obtained normal metrics.

However, when I run validation separately on the trained RT-DETR-l model, all validation metrics return 0 (precision, recall, mAP, etc.).
Interestingly, during the training process itself, the automatic validation performed at the end of each epoch does NOT return 0 values and produces normal metrics.

So the issue only appears when running validation manually after training.

Example behavior:

* Training-time validation for RT-DETR-l → normal metrics
* Manual validation after training → all metrics become 0

I used the same dataset, data.yaml, and validation settings for both models.

Could this be related to:

* RT-DETR postprocessing/NMS differences?
* label parsing during standalone validation?
* confidence threshold handling?
* a bug specific to RT-DETR validation pipeline?

Any help would be appreciated.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @enescicek, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your issue moving quickly, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
Since this appears to be a possible 🐛 bug report affecting manual validation for RT-DETR-l, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
If possible, please include:
- The exact training command and the exact manual validation command used
- The model path used for validation
- Your `data.yaml`
- Package version and environment details
- Any logs showing the difference between training-time validation and manual validation
If this is a custom training ❓ question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
This looks consistent with RT-DETR’s separate [validation flow](https://docs.ultralytics.com/reference/models/rtdetr/val): `postprocess()` filters detections with `self.args.conf`, so if standalone `val` is using a higher `conf` than training-time validation you can end up with all-zero metrics; please retry with `conf=0.001`, and if it still reproduces, paste the exact validation command, full text output, and `yolo checks` output (or `ultralytics.utils.checks.collect_system_info()` from the same Python env) so we can check for a mixed-install/path issue.

**Comment by raimbekovm** [comment]
could you please share a minimum reproducible example (exact train + manual val commands, your `data.yaml`, and the full val output)? @enescicek https://docs.ultralytics.com/help/minimum-reproducible-example/

**Comment by enescicek** [comment]
Thank you for your response.

I managed to solve the issue. However, I still do not understand why it occurred.

When I tried to run validation using the CLI command, all metric values were reported as `0`. However, when I performed the same validation process with a Python script using the same parameters, the results were calculated correctly.

Here is the content of my dataset YAML file:

```yaml
train: /Projects/YoloDatasets/ENES/Arastirma/rt-detr/DATASET/train/images
val: /Projects/YoloDatasets/ENES/Arastirma/rt-detr/DATASET/valid/images
test: /Projects/YoloDatasets/ENES/Arastirma/rt-detr/DATASET/test/images

nc: 5
names:
  [
    "Foramen Mentale",
    "Mandibular canal",
    "Mandibular condyle",
    "Maxillary sinus",
    "Orbita",
  ]
```

These are the parameters I used during training:

```python
model.train(
    data="/Projects/YoloDatasets/ENES/Arastirma/rt-detr/DATASET/dataset.yaml",
    epochs=50,
    imgsz=1280,
    batch=4,
    device=1,
    patience=15,
    fliplr=0.0,
    flipud=0.0,
    task="detect",
    optimizer="AdamW",
    lr0=0.001111,
    momentum=0.9,
    name="Train-rtdetr-l"
)
```

These are the parameters I used during validation with the Python script:

```python
model.val(
    data="/Projects/YoloDatasets/ENES/Arastirma/rt-detr/DATASET/custom.yaml",
    split="test",
    imgsz=1280,
    batch=1,
    device=1,
    task="detect",
    name="Val-rtdetr-l"
)
```

The Python script works correctly, but the equivalent CLI validation command returns `0` for all metrics.

I would appreciate it if you could clarify what might cause this difference between the CLI command and the Python API.

**Comment by raimbekovm** [comment]
there's no inherent difference between the cli and the python api for rt-detr `val`. running the same checkpoint both ways gives identical metrics here, because both go through `RTDETRValidator` rather than the generic detection validator. what differs is how the architecture gets resolved at load time:

- python: loading the checkpoint auto-detects the rt-detr head and switches the model to the `RTDETR` class, which selects `RTDETRValidator`. that head auto-detection from the checkpoint was added in 8.3.166.
- cli: the architecture is guessed from the model filename. a name containing `rtdetr` routes to rt-detr; anything else (yours is `best.pt`) falls back to the generic yolo loader.

so the all-zero cli run matches one specific situation: a checkpoint whose filename doesn't contain `rtdetr`, validated on a version older than 8.3.166. there the cli loads it through the generic path and the standard detection validator runs nms on rt-detr's decoder outputs. those outputs aren't in the format nms expects, almost nothing survives, and every metric reads 0. i reproduced exactly that: feeding an rt-detr model to the detection validator gives `mAP50-95 = 0`, while `RTDETRValidator` gives normal numbers on the same weights. your python script worked because it used the rt-detr path directly and got the right validator.

to confirm which case you hit, share your `yolo checks` output so we can see the version. the fix either way: upgrade to a current release, where the cli auto-detects rt-detr from the checkpoint regardless of filename, or keep `rtdetr` in the weights filename so the cli routes it correctly on older versions. more on the separate [rt-detr val flow](https://docs.ultralytics.com/reference/models/rtdetr/val).
